internal/config/rule.go

```diff
@@ -1,195 +1,126 @@
+		},
+		func(curElement, update *rule.Escalation) error {
+			curElement.NameRaw = update.NameRaw
+			curElement.Condition = update.Condition
+			// TODO: synchronize FallbackForID/Fallback when implemented
+			return nil
+		},
+		func(delElement *rule.Escalation) error {
+			elementRule, ok := r.Rules[delElement.RuleID]
+			if !ok {
+				return fmt.Errorf("rule escalation refers unknown rule %d", delElement.RuleID)
 			}
 
-			for _, escalation := range pendingRule.Escalations {
-				for i, recipient := range escalation.Recipients {
-					recipientLogger := r.logger.With(
-						zap.Int64("id", recipient.ID),
-						zap.Int64("escalation_id", recipient.EscalationID),
-						zap.Int64("channel_id", recipient.ChannelID.Int64),
-						zap.Inline(recipient.Key))
+			delete(elementRule.Escalations, delElement.ID)
+			return nil
+		})
+
+	incrementalApplyPending(
+		r,
+		&r.ruleEscalationRecipients, &r.configChange.ruleEscalationRecipients,
+		func(newElement *rule.EscalationRecipient) error {
+			ok := false
+			if newElement.ContactID.Valid {
+				newElement.Recipient, ok = r.Contacts[newElement.ContactID.Int64]
+			} else if newElement.GroupID.Valid {
+				newElement.Recipient, ok = r.Groups[newElement.GroupID.Int64]
+			} else if newElement.ScheduleID.Valid {
+				newElement.Recipient, ok = r.Schedules[newElement.ScheduleID.Int64]
+			}
+			if !ok {
+				return fmt.Errorf("rule escalation recipient is missing or unknown")
+			}
 
-					if recipient.ContactID.Valid {
-						id := recipient.ContactID.Int64
-						if c := r.Contacts[id]; c != nil {
-							recipient.Recipient = c
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else if recipient.GroupID.Valid {
-						id := recipient.GroupID.Int64
-						if g := r.Groups[id]; g != nil {
-							recipient.Recipient = g
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else if recipient.ScheduleID.Valid {
-						id := recipient.ScheduleID.Int64
-						if s := r.Schedules[id]; s != nil {
-							recipient.Recipient = s
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else {
-						recipientLogger.Warnw("ignoring unknown escalation recipient")
-						escalation.Recipients[i] = nil
-					}
+			ruleFound := false
+			for id, elementRule := range r.Rules {
+				_, ok := elementRule.Escalations[newElement.EscalationID]
+				if ok {
+					newElement.RuleID = id
+					ruleFound = true
+					break
 				}
+			}
+			if !ruleFound {
+				return fmt.Errorf("rule escalation recipient cannot be mapped to a rule")
+			}
 
-				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(r *rule.EscalationRecipient) bool {
-					return r == nil
-				})
+			escalation := r.Rules[newElement.RuleID].Escalations[newElement.EscalationID]
```

Comment:
You can use `r.GetRuleEscalation(newElement.EscalationID)` here instead.

Reply:
Unfortunately, `RuntimeConfig.GetRuleEscalation` only returns a `*rule.Escalation`, but not a reference to the `escalation.Rule`. In particular, the Rule ID is necessary to delete this `EscalationRecipient` again, as there is no link to `Escalations` at the `RuntimeConfig`.

https://github.com/Icinga/icinga-notifications/blob/767e4b8ef142c6e0bfd11573ace18576f6811906/internal/config/rule.go#L110-L125

Reply:
> In particular, the Rule ID is necessary to delete this `EscalationRecipient` again, as there is no link to `Escalations` at the `RuntimeConfig`.

If `RuntimeConfig.GetRuleEscalation` does not return nil, you can simply access the rule ID via `escalation.RuleID`..

Reply:
IMHO, adding the additional field `RuleID` to `EscalationRecipient` is not even necessary as a recipient has a reference to its escalation and the escalation to its rule.

Reply:
Valid point. Now I am going to directly refer `*Escalation`, easing both the `createFn` and `delteFn`. Thanks :)